drivers/timer/nrf_rtc_timalarm.c

```diff
@@ -0,0 +1,459 @@
+	 * passed (causing immediate callback) if uncounted ticks have
+	 * occurred.
+	 *
+	 * A positive tick schedules for the requested number of ticks
+	 * after the last announced tick.  This too may be in the
+	 * past.
+	 */
+
+	/* Explain why we need to do this. */
+	(void)k_alarm_cancel(&tick_alarm);
+
+	if (ticks == K_FOREVER) {
+		/* "no future timer interrupts are expected or
+		 * required"
+		 *
+		 * At this point we have no obligation to maintain the
+		 * tick clock; consequently in tick calculation code
+		 * we are entitled to act as though last_tick_sc is
+		 * within a half 32-bit span of the current time, an
+		 * invariant maintained by MAX_TICKS.
+		 */
+		return;
+	}
+
+	u32_t nts;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	u32_t ntd = (sysclock_get_32() - last_tick_sc) / SC_PER_TICK;
+
+	if (ntd == 0) {
+		/* Can't announce zero ticks */
+		ntd = 1;
+	}
+	if (ticks > 0) {
+		/* Kernel wants a timeout with a positive offset
+		 * relative to the last announced tick.
+		 */
+		ntd = MAX(ntd, MIN(ticks, MAX_TICKS));
+	}
+	next_tick_delta = ntd;
+	next_tick_sc = last_tick_sc + next_tick_delta * SC_PER_TICK;
+	nts = next_tick_sc;
+	k_spin_unlock(&lock, key);
+
+	(void)k_alarm_schedule(&tick_alarm, nts, 0);
+#endif
+}
+
+u32_t z_clock_elapsed(void)
+{
+	u32_t rv = 0;
+#ifdef CONFIG_TICKLESS_KERNEL
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	rv = (sysclock_get_32() - last_tick_sc) / SC_PER_TICK;
+
+	k_spin_unlock(&lock, key);
+#endif
+	return rv;
+}
+
+/* Implement z_alarm_update_deadline. */
+static void alarm_update_deadline_(u32_t now)
+{
+	u32_t compare = now;
+	u32_t deadline = 0;
+	int rc = k_alarm_next_deadline_(&deadline);
+
+	if (rc < 0) {
+		/* No unscheduled alarms, use maximum delay */
+		compare += COUNTER_MAX_DELTA;
+	} else if (rc > 0) {
+		/* Next event at deadline.  If that's now or in the
+		 * past target a compare of now; otherwise queue an
+		 * event at that deadline or the maximum delta from
+		 * now whichever's sooner.
+		 */
+		u32_t delay = deadline - now;
+
+		if ((s32_t)delay > 0) {
+			if (delay < COUNTER_MAX_DELTA) {
+				compare += delay;
+			} else {
+				compare += COUNTER_MAX_DELTA;
+			}
+		} else {
+			/* Alarm is (past) due, fire ASAP */
+		}
+	} else {
+		/* Something is ready, fire ASAP */
+	}
+
+	/* compare is at most COUNTER_MAX_DELTA past now.  The RTC
+	 * counter should not have advanced more than
+	 * COUNTER_MAX_DELTA - 2 ticks past now.
+	 *
+	 * If the next event is already due and we haven't yet cleared
+	 * the last event, leave it enabled so we'll re-enter the
+	 * FLIH.
+	 */
+	if ((compare == now)
+	    && RTC->EVENTS_COMPARE[0]) {
+		return;
+	}
+
+	/* Make sure compare is at least COUNTER_MIN_DELTA past now,
+	 * then clear the COMPARE event and set the compare value.
+	 */
+	if (COUNTER_MIN_DELTA > (compare - now)) {
+		compare = now + COUNTER_MIN_DELTA;
+	}
+	RTC->EVENTS_COMPARE[0] = 0;
+	RTC->CC[0] = compare;
```

Comment:
In this case the problem would only arise if `COUNTER` incremented twice after `now` was read, with one of those times being after the check that `COMPARE` was clear.  In that case we'd be at most 2 ticks late.  The alternative approach you cite is extremely complicated and I strongly expect the logic required to implement it would have a worse impact than an extremely rare delay in an alarm that's already subject to latency from arbitrary interrupts.

Reply:
On the other hand current approach has following consequences:
- enforcing high priority of RTC interrupt because access to RTC registers must not be interrupted (`COUNTER_MIN_DELTA`)
- User callbacks are also executed in RTC interrupt context. That is risky and may impact system performance. Especially, if application requires other RT operations in high priority interrupts. For unaware users that may have even more consequences (e.g. using printk from timeout callback leading to bluetooth disconnections).
- delaying expiration of late alarms by 4 cycles. 

RTC interrupt should rather have low/lowest priority with algorithm that handles late setting. 

>  The alternative approach you cite is extremely complicated

well, it's like 15-20 lines of code, if that would allow lowering interrupt priority i would go for that.

Reply:
RTC registers are protected by a spinlock (disabling interrupts); the priority of the RTC interrupt is not changed by this implementation.  Nor is the ability for applications to provide callbacks that are executed from within the ISR.  The delay exists as an edge case and is due to limitations in the Nordic hardware; by using a dedicated peripheral rather than channel on peripheral, could be eliminated by the [technique you suggested](https://github.com/zephyrproject-rtos/zephyr/pull/16398#discussion_r288082113).

If something like k_alarm is to be a replacement for the existing timer infrastructure your concerns should be considered in the context of defining scope and functional requirements that can be used as the basis of target-specific implementation.